frontend/app/src/next/components/waterfall/waterfall.tsx

```diff
@@ -0,0 +1,467 @@
+        );
+
+        taskHasChildrenMap.set(task.metadata.id, hasChildren);
+
+        // Record task depth from API or calculate
+        taskDepthMap.set(task.metadata.id, task.depth);
+
+        // Record parent-child relationship
+        if (task.depth == 0) {
+          rootTasks.push(task.metadata.id);
+        } else if (task.parentTaskExternalId) {
+          if (!taskParentMap.has(task.parentTaskExternalId)) {
+            taskParentMap.set(task.parentTaskExternalId, []);
+          }
+          const children = taskParentMap.get(task.parentTaskExternalId);
+          if (children) {
+            children.push(task.metadata.id);
+          }
+        }
+
+        // Initialize path
+        taskPathMap.set(task.metadata.id, [task.metadata.id]);
+      }
+    });
+
+    // Second pass: build complete paths
+    taskMap.forEach((task) => {
+      if (
+        task.metadata?.id &&
+        task.parentTaskExternalId &&
+        taskPathMap.has(task.parentTaskExternalId)
+      ) {
+        const parentPath = taskPathMap.get(task.parentTaskExternalId);
+        if (parentPath) {
+          taskPathMap.set(task.metadata.id, [...parentPath, task.metadata.id]);
+        }
+      }
+    });
+
+    // Auto-expand first set of root tasks with children
+    if (!autoExpandedInitially && taskData) {
+      // Find root tasks with children and expand them
+      const rootTasksWithChildren = rootTasks.filter((id) =>
+        taskHasChildrenMap.get(id),
+      );
+      if (rootTasksWithChildren.length > 0) {
+        // Only expand the first time when data is available
+        setTimeout(() => {
+          const newExpandedTasks = new Set(expandedTasks);
+          rootTasksWithChildren.forEach((id) => {
+            newExpandedTasks.add(id);
+          });
+          setExpandedTasks(newExpandedTasks);
+          setAutoExpandedInitially(true);
+        }, 0);
+      }
+    }
+
+    // Determine which tasks should be visible based on expanded state
+    const visibleTasks = new Set<string>();
+
+    // Always add root tasks to visible tasks
+    rootTasks.forEach((id) => {
+      visibleTasks.add(id);
+    });
+
+    // Add children of expanded tasks
+    expandedTasks.forEach((expandedId) => {
+      const children = taskParentMap.get(expandedId) || [];
+      children.forEach((childId) => {
+        visibleTasks.add(childId);
+      });
+    });
+
+    // If there are no visible tasks, return early to avoid errors
+    if (visibleTasks.size === 0) {
+      return { data: [], taskPathMap: new Map() };
+    }
+
+    // Find the global minimum queuedAt time among visible tasks
+    let globalMinQueuedAt = Number.MAX_SAFE_INTEGER;
+    visibleTasks.forEach((id) => {
+      const task = taskMap.get(id);
+      if (task && task.queuedAt) {
+        const queuedTime = new Date(task.queuedAt).getTime();
+        if (queuedTime < globalMinQueuedAt) {
+          globalMinQueuedAt = queuedTime;
+        }
+      }
+    });
```

Comment:
this is such a Go way of writing this 😅  

I don't feel strongly here, but in general I like the functional style a lot when we can use it, less prone to bugs IMO (untested code below)

```javascript
[...visibleTasks].reduce(
  (acc, id) =>
    Math.min(new Date(taskMap.get(id)?.queuedAt || Date.now()).getTime(), acc),
  Number.MAX_SAFE_INTEGER,
);
```

I also think this might be a little simpler if `visibleTasks` is a list and I figure performance isn't really an issue, but not a big deal either